### ISSUE TITLE
Incorrect configuration error for no-param-reassign

### ISSUE THREAD

**Issue by OliverSieweke** [eslint/eslint#11837]
**Tell us about your environment**

* **ESLint Version:** 5.16.0
* **Node Version:** 10.16.0 (current LTS)
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

Minimal working example for `.eslintrc.json`
```js
{
    "rules": {
        "no-param-reassign": [
            "warn",
            {
                "props": false,
                "ignorePropertyModificationsFor": []
            }
        ]
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

Empty `test.js`:
```js

```

```bash
eslint test.js
```

**What did you expect to happen?**
No error.

**What actually happened? Please include the actual, raw output from ESLint.**
Invalid configuration:
```
Error: .../.eslintrc.json:
        Configuration for rule "no-param-reassign" is invalid:
        Value {"props":false,"ignorePropertyModificationsFor":[]} should NOT have additional properties.
        Value false should be equal to one of the allowed values.
        Value {"props":false,"ignorePropertyModificationsFor":[]} should match exactly one schema in oneOf.
```

**Comments**

This only happens when the `props` option is explicitly set to `false` (which is also its default value).

I'm happy to submit a pull request if this is recognized as a bug.

**Comment by aladdin-add** [comment]
thanks for reporting! seems a bug to me, labelled accept.

**Comment by platinumazure** [comment]
Just for my understanding, what does `ignorePropertyModificationsFor` do when `props` is `false`?

**Comment by OliverSieweke** [comment]
`ignorePropertyModificationsFor` does not have any effect when `props` is `false` as it only intervenes in the following conditional: 
```js
if (props && isModifyingProp(reference) && ignoredPropertyAssignmentsFor.indexOf(identifier.name) === -1)`)
```

From that angle, it does make some sense that `props` should not be set to `false` when the `ignorePropertyModificationsFor` property exists. 

I find it more confusing to get an error when setting an option back to its default value, especially because this is not how ESLint behaves for other rules.

I don't feel strongly about this and one could probably have a more general discussion about it. At the moment the proposed change would lead to more consistency with other rules, where one option only makes sense when coupled with another one. Some collected examples:
- `new-cap`: `newIsCapExceptions` only makes sense with `newIsCap` set to `true`, but having both options with `newIsCap` set to `false` does not lead to a schema error.
- `spaced-comment`: `exceptions` only makes sense with the rule set to `always`, but adding exceptions with the rule set to `never` does not lead to a schema error.
- `camelcase`: `allow` only makes sense when the following options are not set simultaneously `{properties: "never", ignoreDestructuring: true}`. Again no schema error arises if that is not the case.
- `padded-blocks`: `allowSingleLineBlocks` only makes sense with the rule set to `always`, but having the property with the rule set to `never` does not cause any schema error.
- `lines-between-class-members`: `exceptAfterSingleLine` only makes sense with the rule set to `always` but again this is not enforced in the schema.
- `id-length`: `exceptions` only makes sense with `min` or `max` set but is also allowed when none of them is set.

**Comment by platinumazure** [comment]
I need to fully review your latest comment, but I'm noting (more for posterity) that a lot of the problem might have to do with the vagueness of the error message you received. If the message specifically said that the property in question isn't valid and should be removed, this would probably be less of an issue (although you may disagree).

**Comment by kaicataldo** [comment]
I'm not personally convinced this should be labeled a bug. While I think the generic configuration error isn't particularly helpful, I do think the end user should be aware that `ignorePropertyModificationsFor` doesn't do anything when `props` is `false`.

**Comment by OliverSieweke** [comment]
@kaicataldo No worries, this seems to require some more discussion.
@platinumazure Given that the error message specifies the rule that is concerned, it's not that huge an issue to figure out what went wrong, although a more precise error message would definitely be an improvement!

Some more thoughts on the issue:

- If people think that it is preferable for the user to be aware of "inactive" options, then this logic should be applied across all rules including those listed at the end of my [previous comment](https://github.com/eslint/eslint/issues/11837#issuecomment-502759044) (`new-cap`, `spaced-comment`, `camelcase`, `padded-blocks`, `id-length`). This reasoning could also be extended to entire rules: if a rule is turned off, even though the user specified options for it, then he should be made aware of it.
- There is definitely good case to be made in favor of the above. Some possibly negative consequences to consider:
  - It would lead to some configurations being invalid, although all options taken separately contain valid values. Without more precise error messages or pointers in the docs this could be a bit confusing and is what triggered this issue to start with.
  - I like to set all options explicitly, for instance, so that when I realize that a rule is not quite working as intended, I can quickly go to the config file and correct it, add exceptions etc. without necessarily having to go through the docs. This kind of approach wouldn't work anymore if it is not possible to set "inactive" options.

- Regardless of the best approach, the configuration schemas for "inactive" options should probably be handled consistently across rules, which isn't quite the case at the moment.
